BC-10946 External person invitation and Room Rights

joysenthil opened the pull request:
# Description

<!--
  This is a template to add as much information as possible to the pull request, to help reviewer and as a checklist for you. Points to remember are set in the comments, please read and keep them in mind:
    - Code should be self-explanatory and share your knowledge with others
    - Document code that is not self-explanatory
    - Think about bugs and keep security in mind
    - Write tests (Unit and Integration), also for error cases
    - Main logic should be hidden behind the api, never trust the client
    - Visible changes should be discussed with the UX-Team from the beginning of development; they also have to accept them at the end
    - Keep the changelog up-to-date
    - Leave the code cleaner than you found it. Remove unnecessary lines. Listen to the linter.
-->

## Links to Tickets or other pull requests
BC-10946
<!-- related-prs-and-tickets-start -->
<!-- related-prs-and-tickets-end -->

## Screenshots of UI changes

<!--
  only needed for visual changes
  If visual changes exist, work together with UI/UX from beginning/ping UX with final PR
-->

## Approval for review

- [ ] DEV: If the API or client code was changed, all necessary code generation or synchronization steps were completed and tested.
- [ ] QA: In addition to review, the code has been manually tested (if manual testing is possible)
- [ ] All points were discussed with the ticket creator, support-team or product owner. The code upholds all quality guidelines from the PR-template.

Commit:
BC-10946 Initial commit with External person invitation and acceptance

Commit:
BC-10946 Expternal person invitation without confirmation feature added

Commit:
BC-10946 Negative checking added to the feature tests

Commit:
BC-10946 : added additional assertions

Commit:
Merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

Commit:
Merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

Commit:
BC-10946: added negative assertions to the feature file.

Commit:
BC-10946 Test assertions activated

UzaeirKhan committed:
BC-10946: Merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: following files has been updated:

- *.feature: update step for navigating to room
- *.js: update method
- runSchoolApi.js: update expert -> externalPerson

UzaeirKhan committed:
BC-10946: Merge branch 'main' into 'BC-10946-EP-invitatoin-links-and-Room-Rights'

UzaeirKhan committed:
BC-10946: merge branch 'main' into 'BC-10946-EP-invitatoin-links-and-Room-Rights'

UzaeirKhan committed:
BC-10946: following files has been updated:

- tests(rooms): rename feature file typo
  - roomInvitationViaLinkExternalPeronWithConfirmation.feature -> roomInvitationViaLinkExternalPersonWithConfirmation.feature

- tests(rooms): improve Gherkin titles/description

- auth(login.helper): robust extern detection
  - Replace substring check with token-based regex: /(?:^|_)extern(?:_|$)/ to avoid matching “externalPerson”.

- pages(rooms): add selectors for new FAB actions
  - [data-testid="fab-select-from-directory"]
  - [data-testid="fab-add-external-person"]

- pages(courses_board): adjust selectors/usage to match current UI (minor)

- scripts(runSchoolApi): minor alignment with login changes (logging/flow)

UzaeirKhan committed:
BC-10946: rename feature file

UzaeirKhan committed:
BC-10946: removed unused steps and methods

UzaeirKhan committed:
BC-10946: removed staging tags from feature file

UzaeirKhan committed:
BC-10946: named selector in the openThreeDotMenuForRoom method

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithConfirmation.feature:
```diff
@@ -0,0 +1,76 @@
+@prio_0_dev
+Feature: Rooms - Room invitations via link with confirmation for external persons
+
+    As a room owner, I create an invitation link that requires confirmation; external persons request access via the link and must be approved before joining.
```
this can be -> _As a room owner, I want to create an invitation link that requires approval, so that external users can request access and only join after I approve them._

UzaeirKhan replied:
done

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithConfirmation.feature:
```diff
@@ -0,0 +1,76 @@
+        When I check the Checkbox to valid for external persons
+        When I save the invitation link
+        Then I see the Link URL in the Modal
+        When I remember the invitation link URL in the Modal
```
this step should be -> _When I copy the invitation link URL in the Modal_ , instead of 'remember' ?

csp175 replied:
When I copy the invitation link URL in the modal , instead of 'remember' ?

UzaeirKhan replied:
done

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithConfirmation.feature:
```diff
@@ -0,0 +1,76 @@
+
+        # external person uses the invitation link to join the room
+        Given I am logged in as a '<external_person_1>' at '<namespace>'
+        When I use the remembered invitation link URL
```
also here ->  _When I open the copied invitation link URL_

UzaeirKhan replied:
done

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithConfirmation.feature:
```diff
@@ -0,0 +1,76 @@
+        When I select the three dot menu action 'room-members'
+        Then I see the page Edit participants of room '<room_name>'
+        Then I see teacher '<external_person_last_name>' not visible in the table
+
```
line space here

UzaeirKhan replied:
done

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithConfirmation.feature:
```diff
@@ -0,0 +1,76 @@
+        When I click on tab Confirmations
+        Then I see user '<external_person_last_name>' in the confirmations table
+        When I click on button Three Dot Menu in Confirmations table for user '<external_person_last_name>'
+        When I click on confirm button in the three dot menu
```
should be ->  _When I click on button Confirm in the three dot menu_

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithoutConfirmation.feature:
```diff
@@ -0,0 +1,71 @@
+
+    Scenario Outline: Room Owner creates no-confirmation link; external person joins, rights are limited, leaves; owner verifies removal
+
+        # pre-condition: users (teacher & external person) logged in
```
same here too -> _users (teacher & external person) accounts are created._

UzaeirKhan replied:
same as in other scenario

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithoutConfirmation.feature:
```diff
@@ -0,0 +1,71 @@
+
+        # external person uses the no-confirmation invitation link to join the room
+        Given I am logged in as a '<external_person_1>' at '<namespace>'
+        When I use the remembered invitation link URL
```
Same changes needed as in the above other feature file.

UzaeirKhan replied:
done

atlldwp commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithoutConfirmation.feature:
```diff
@@ -0,0 +1,71 @@
+        When I click on three dot menu in room page
+        When I select the three dot menu action 'leave-room'
+        Then I see dialog box to leave the room
+        Then I click on button 'Confirm' to leave the room
```
Here dialog confirmation is parameterized? because in other scenario this kind of step requires just click on the button without any param for the button.

UzaeirKhan replied:
Auto-join scenario only allows leaving (no confirmation needed), so dialog step is parameterized unlike the confirmation-required flow.

atlldwp committed:
Merge branch 'main' of https://github.com/hpi-schul-cloud/e2e-system-tests into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10496: adjust feature for staging environment

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: Merge branch 'BC-10946-EP-invitatoin-links-and-Room-Rights' of github.com:hpi-schul-cloud/e2e-system-tests into BC-10946-EP-invitatoin-links-and-Room-Rights

atlldwp reviewed (changes requested):
see comments!

UzaeirKhan committed:
BC-10946: revert the changes in feature files for additional steps

UzaeirKhan committed:
BC-10946: update after review part 1

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: merge branch 'main' into 'BC-10946-EP-invitatoin-links-and-Room-Rights'

UzaeirKhan commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithConfirmation.feature:
```diff
@@ -0,0 +1,76 @@
+
+    Scenario Outline: Room Owner creates confirmation required link, external person applies, owner approves, access is granted
+
+        # pre-condition: users (teacher & external person) logged in
```
for school API it makes sense but for staging users we are not creating them so better to mention users are logged in

UzaeirKhan commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithConfirmation.feature:
```diff
@@ -0,0 +1,76 @@
+        When I click on tab Invitations
+        When I click on the fab button to create an invitation link
+        Then I see the modal Create Invitation Link
+        When I enter '<invitation_description>' into the Invitation Link Description field
```
done

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithoutConfirmation.feature:
```diff
@@ -0,0 +1,71 @@
+        When I check the Checkbox to valid for external persons
+        When I save the invitation link
+        Then I see the Link URL in the Modal
+        When I remember the invitation link URL in the Modal
```
done

UzaeirKhan commented on cypress/e2e/rooms/roomInvitationLinks/roomInvitationViaLinkExternalPersonWithoutConfirmation.feature:
```diff
@@ -0,0 +1,71 @@
+        Given I am logged in as a '<external_person_1>' at '<namespace>'
+        Given I am logged in as a '<teacher_1>' at '<namespace>'
+
+        # pre-condition: teacher creating a new room
```
done

UzaeirKhan committed:
BC-10946: following files has been updated:

- *.feature: update steps and added staging tag and example table for relevant features
- */pages/*.js: added new methods and update method names after step changes
- */*.spec.js: changed common steps location and added new steps

UzaeirKhan committed:
BC-10946: revert staging feature file changes

UzaeirKhan committed:
BC-10946: renamed step in feature and step definition

UzaeirKhan committed:
BC-10946: Revert admin page object file and remove unused steps/selectors

UzaeirKhan committed:
BC-10946: refactored after copilot review

UzaeirKhan commented on cypress/support/pages/courses_board/pageCoursesBoard.js:
```diff
@@ -196,9 +196,12 @@ class Board {
+		cy.get(`[data-testid="kebab-menu-action-${kebabMenuAction.toLowerCase()}"]`).click({
+			force: true,
+		});
+		cy.get(`[data-testid="kebab-menu-action-${kebabMenuAction.toLowerCase()}"]`).should(
```
yes, it's needed to prevent timing issues

UzaeirKhan committed:
BC-10946: rephrase step to align with guidelines

UzaeirKhan committed:
BC-10946: update example tables in feature files

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

atlldwp committed:
Merge branch 'main' of https://github.com/hpi-schul-cloud/e2e-system-tests into BC-10946-EP-invitatoin-links-and-Room-Rights

atlldwp committed:
Merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: removed duplicated selectors in pageRooms.js file

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: merge branch 'BC-10946-EP-invitatoin-links-and-Room-Rights' of github.com:hpi-schul-cloud/e2e-system-tests into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: prevent timing issues in menu action button method

UzaeirKhan committed:
BC-10946: fixed menu closes before click issue for external person

UzaeirKhan committed:
BC-10946: refactored focus issue in three dot menu

UzaeirKhan committed:
BC-10946: refactored menu action method

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: corrected focus/blue issue in openThreeDotMenuForRoom method

UzaeirKhan committed:
BC-10946: merge branch 'main' into BC-10946-EP-invitatoin-links-and-Room-Rights

UzaeirKhan committed:
BC-10946: removed commented code
